### DISCUSSION TITLE
Code Block

### DISCUSSION THREAD

**Discussion by wartburggraaf** [nl-design-system/backlog#328] [Component suggestions]
<img width="480" alt="visuele weergave van een Code block component" src="https://github.com/user-attachments/assets/7945ee44-c47c-4604-8d66-8c17e2bd618d">

We willen graag meer leren over de **Code Block** component. In welke vorm, of vormen, wordt deze momenteel ingezet en toegepast binnen jouw organisatie?

Deel hieronder een link naar Figma, Storybook, Github of screenshots. Met deze informatie kunnen we bewijzen dat de component onderdeel moet worden van de NL Design System component library.

Daarnaast wordt het delen van onderzoek gewaardeerd. Dus heb je zelf onderzoek gedaan en zijn er inzichten voor dit component? Dan horen we dit ook graag!

[Link naar bericht in Slack](https://codefornl.slack.com/archives/C7KDQRXMZ/p1725369369391399)

---

## Naam
Code Block

## Doel

Blok met één of meerdere regels computercode.

## Gerelateerde componenten

[Code](https://github.com/orgs/nl-design-system/discussions/327)

## Richtlijnen

Acceptatiecriteria: [NL Design System Acceptatiecriteria voor Code Block](https://nldesignsystem.nl/code-block)

## Onderzoek

[Poll over gedrag van Code Block kleurstelling bij Light of Dark mode - Mastodon](https://fosstodon.org/@BryceWrayTX/113629960883353006)

**Comment by jeffreylauwers** [comment]
## Utrecht

- [Storybook](https://nl-design-system.github.io/utrecht/storybook/?path=/docs/css_css-code-block--docs)
- [GitHub](https://github.com/nl-design-system/utrecht/tree/main/components/code-block)

---

### Screenshots / Varianten

**Storybook**

Default
Markup with the `<pre>` and `<code> `elements. Styling via `utrecht-code-block` class name.
Only using the `<pre>` element for markup is semantically ambiguous, it is clearer to use `<pre><code>...</code></pre>`.

![Scherm­afbeelding 2024-09-04 om 14 03 18](https://github.com/user-attachments/assets/626cbda4-15ba-4642-bba8-8ca33a44a0b2)

---

### Notities

- Via 'Element' property kan er gekozen worden voor `div` of `pre`.
- Lichtgrijze achtergrondkleur.

**Comment by jeffreylauwers** [comment]
## Atlassian

[Documentatie website](https://atlassian.design/components/code/code-block/examples)

---

### Screenshots / Varianten

**Documentatie website**

Default
A code block highlights an entire block of code and keeps the formatting.

![Scherm­afbeelding 2024-09-04 om 14 05 46](https://github.com/user-attachments/assets/41b88ba7-810e-4d8a-b88c-23d1219cae85)

Line numbers
You can display a code block with or without line numbers. Line numbers are included by default.

![Scherm­afbeelding 2024-09-04 om 14 06 42](https://github.com/user-attachments/assets/e3d7a714-cb04-47ea-87a5-8fb791785c8f)

---

### Notities

- Onder de naam 'Code block'
- Lichtgrijze, transparent, achtergrond kleur. Subtiel afgeronde hoekjes.
- Veel varianten beschikbaar. Naast bovenstaande ook de mogelijkheid om maar bepaalde regelnummers te tonen, regels van een 'highlight' te voorzien en om lange regels code te wrappen.

**Comment by jeffreylauwers** [comment]
## IBM

[Documentatie website](https://carbondesignsystem.com/components/code-snippet/usage/)

---

### Screenshots / Varianten

**Documentatie website**

Code snippet - 'Multiline'
Code snippets are strings or small blocks of reusable code that can be copied and inserted in a code file.

![Scherm­afbeelding 2024-09-04 om 14 12 24](https://github.com/user-attachments/assets/d41e1efb-57a9-476b-8e3f-2444d7c4c9f2)

---

### Notities

- Onder de naam 'Code snippet' als variant 'Multiline'.
- Lichtgrijze achtergrond kleur. Geen afgeronde hoekjes.
- Optie rechtsboven om code te kopieren.
- Optie rechtsonder om meer code te tonen.

**Comment by jeffreylauwers** [comment]
## GitHub

[Documentatie website](https://primer.style/foundations/typography#code)

---

### Screenshots / Varianten

**Documentatie website**

![Scherm­afbeelding 2024-09-04 om 14 15 29](https://github.com/user-attachments/assets/afc257fc-aa25-4437-bb68-3acaf842bdfc)

---

### Notities

- Onder de naam 'Code block' beschikbaar als typografisch element
- Zie ook de code weergave die binnen Github zelf gebruikt wordt (bijvoorbeeld op de [React pagina](https://primer.style/guides/react)). Daar wordt een Code block net even anders weergegeven (lichtgrijze achtergrond, kleur en kopier optie). Dit is niet terug te vinden als component.

**Comment by jeffreylauwers** [comment]
## NL Design System

[Documentatie website](https://www.nldesignsystem.nl/richtlijnen/formulieren/descriptions/koppelen)

---

### Screenshots / Varianten

**Documentatie website**

Op de documentatie website van NL Design Sytem gebruiken we de Code component van Utrecht.

Default
![Scherm­afbeelding 2024-09-04 om 14 20 07](https://github.com/user-attachments/assets/c5e8289e-651d-4295-847b-61c7f99c4f6d)

Als onderdeel van een voorbeeld. Met daarbij een 'Toon/Verberg' optie en de mogelijkheid om de code te kopieren.

![Scherm­afbeelding 2024-09-04 om 14 20 13](https://github.com/user-attachments/assets/043a7133-60e8-4eab-82e4-dd47b845deca)

---

Code heeft onderscheidende kleuren.

**Comment by jeffreylauwers** [comment]
## ✨ Dit component is nu Help Wanted ✨

Help je mee hem door de Community stappen te krijgen?
[Code block op het Community bord](https://github.com/orgs/nl-design-system/projects/29/views/1?filterQuery=code+%23329)

**Comment by jeffreylauwers** [comment]
## ✨ Dit component is nu Community ✨

Help je mee hem door de Candidate stappen te krijgen?
[Code Block op het Candidate bord](https://github.com/orgs/nl-design-system/projects/32/views/1?filterQuery=Code+block)

**Comment by rianrietveld** [comment]
@jeffreylauwers 
Leesvoer van Eric Meyer: https://meyerweb.com/eric/thoughts/2022/12/29/styling-a-pre-that-contains-a-code/

**Reply by jeffreylauwers** [reply]
He :has() a point... sorry 😉.

@matijs en @Robbert wat denken jullie? Slim om dit voor de Candidate component te implementeren?

**Comment by hidde** [comment]
In sommige code blocks is horizontale scroll nodig omdat niet alle code binnen de beschikbare ruimte past. 

Bij die code blocks is het voor [2.1.1 Toetsenbord](https://nldesignsystem.nl/wcag/2.1.1) essentieel dat dit scrollen ook met alleen een toetsenbord kan worden gedaan. Dat kan alleen als het element focusbaar is.

Op dit moment maakt alleen Firefox elementen met overflow focusbaar, [Chromium werkt eraan](https://developer.chrome.com/blog/keyboard-focusable-scrollers) en van Safari/Webkit is de status onbekend.

Er is een oplossing, namelijk het toevoegen van `tabindex=0` aan code blocks, zo wordt het element focusbaar zonder de tabvolgorde te beïnvloeden. Maar zomaar een focusbaar element is onverwacht gedrag voor gebruikers, voorstel is om dus `role=region` toe te voegen alsmede een toegankelijke naam. Op de NL Design System site hebben we dit al gedaan (zie [#1726](https://github.com/nl-design-system/documentatie/pull/1726/) en https://github.com/nl-design-system/documentatie/pull/1735/).

Liefst ook met een observer die zorgt dat de tabindex er alleen is wanneer het element daadwerkelijk scrollbaar is (hetgeen kan veranderen bij resize, vandaar de noodzaak voor iets dat kan observeren).

De combo van observed tabindex, region role en acc name doen ze bij o.a. Astro docs en bij GOV.UK Design System.

Zie ook: 

- [Keyboard only scrolling areas](https://adrianroselli.com/2022/06/keyboard-only-scrolling-areas.html) van Adrian Roselli 
- [Keyboard focusable scrollers](https://developer.chrome.com/blog/keyboard-focusable-scrollers) van Chrome blog (zou er op moment van schrijven vanaf Chromium 132 in moeten zitten; getest met Canary 132, waar dat op dit moment inderdaad zo is)

**Comment by jeffreylauwers** [comment]
## ✨ Dit component is nu Candidate ✨

Help je mee hem door de Hall of Fame stappen te krijgen?
[Code Block op het Hall of Fame bord](https://github.com/orgs/nl-design-system/projects/30/views/1?filterQuery=code+block)

**Comment by emma-koole** [comment]
De functionaliteit ‘kopiëren’ is gewenst in het Code Block.
